Documentation improvements around recursion

lstwn opened the pull request:
Hi 👋

As I've recently said on Discord, I have found some areas where the documentation around recursion/iteration could be improved, while digging into these topics for my Master's thesis. Here's what I've changed:

1. The `Circuit::iterate` code example panicked if executed (which it wasn't previously upon doc tests).
2. The `ChildCircuit<P>::recursive()` code example is a bit more friendly towards copying and then playing around with (using arrays instead of `Vec`s leads to compile time checks if the lengths of the inputs and expected outputs don't match anymore).
3. I added a tutorial dedicated to fixed-point computations by showing how to compute a transitive closure. I suspect this is a problem readers may be familiar with.

Hope you find it helpful!

What I would be interested in beyond this: How would you resolve the issue mentioned in the new tutorial about not dealing with a fixed point in case of cyclic graphs. For instance, is there a way to exclude "columns" (here the cumulated weight and hop count) from the consideration of what a fixed-point entails? Or is there a way to keep track of discovered paths and then abort the search if a new path is just an extension of a previously discovered path, through e.g. DBSP's support of iteration with arbitrary conditions?

lstwn committed:
Demonstrate panic of doc example code through minimal changes

Signed-off-by: Leo Stewen <lstwn@mailbox.org>

lstwn committed:
Fix and improve doc example code for Circuit::iterate

Additionally, ensure that the example code has an assertion at all to prevent
the example from being broken in the future.

Signed-off-by: Leo Stewen <lstwn@mailbox.org>

lstwn committed:
Improve ChildCircuit<P>::recursive() doc example code

Next to a clearer variable naming, the example code is a bit more inviting
to play around with if a user copies it, by having the array lengths be
statically kept in sync through the `STEP` constant.

Signed-off-by: Leo Stewen <lstwn@mailbox.org>

lstwn committed:
Add tutorial 10 on fixed-point computations

- Introduced a new example `tutorial10.rs` demonstrating fixed-point
  computations by computing a transitive closure for a graph.
- Updated `Cargo.toml` to include the new example.
- Expanded the tutorial documentation with a section on fixed-point
  computations, including explanations and references to tutorial 10.

Signed-off-by: Leo Stewen <lstwn@mailbox.org>

blp commented:
I :heart: seeing PRs from outside the core team!

ryzhyk commented:
Amazing, thanks for the contribution! I'll do a careful review shortly.

ryzhyk reviewed (commented):
Looks great, thanks again for the contribution. It would be nice if you could include a fix to get the example program to terminate as part of the tutorial. The most straightforward solution is to filter out newly discovered paths that are longer than a previously found path between the same nodes. The simplest way to do this is probably to index the output by (source, destination) pair and aggregate the result to only keep the tuple with the smallest weight.

Note that in general recursive programs with aggregates are not guaranteed to converge even when there exists a finite solution, because they are not guaranteed to be monotonic, but this one should be well-behaved.

lstwn committed:
Add tutorial 11 example to demonstrate termination of recursive
computation with cycles

The example demonstrates how to compute the transitive closure using
recursive queries with the `Min` aggregation operator to ensure
termination in the presence of cycles. This addresses the feedback on the PR.

Signed-off-by: Leo Stewen <lstwn@mailbox.org>

lstwn commented:
Thanks for the feedback! I've addressed it and extended the tutorial. You may want to double check the last paragraph about the convergence of recursive programs with aggregates. It may benefit from a concrete example about a case where convergence is _not_ guaranteed, I think, but I fail to come up with one from the top of my head.

ryzhyk commented on crates/dbsp/src/tutorial.rs:
````diff
@@ -1950,6 +1951,298 @@
+//! ```
+//!
+//! Finally, we point out that introducing a cycle to the graph prevents
+//! this fixed-point computation to stop terminating because then
````
```suggestion
//! this fixed-point computation from terminating because then
```

ryzhyk commented on crates/dbsp/src/tutorial.rs:
```diff
@@ -1950,6 +1951,298 @@
+//! [`Min`](`crate::operator::dynamic::aggregate::Min`) aggregation operator
+//! to only retain the shortest path for each pair.
+//! Aggregation requires to index the stream, so there are more code changes
+//! required than shown here. You can find the full code in `tutorial11.rs` but
```
> Aggregation requires to index the stream, so there are more code changes required than shown here

Not sure I understand this sentence. I think the code below is complete, including indexing the stream?

lstwn replied:
Not quite if compared to tutorial 10:

1. The `len_1` stream requires a `map_index` instead of a `map`. 
2. The `expected_outputs` are now `indexed_zset`s instead of `zset`s.

Both changes are hidden in the docs. But as these are quite minor, I can also remove this hint. Whatever you prefer.

ryzhyk commented on crates/dbsp/src/tutorial.rs:
```diff
@@ -1950,6 +1951,298 @@
+//! recursive queries with aggregates are not guaranteed to converge to the
+//! optimum of the aggregation function (here, the minimum function),
+//! even though there exists a finite solution.
+//! If the problem at hand is not _monotonic_, the computed fixed-point may
+//! not be the optimal solution to the aggregation function.
```
```suggestion
```

ryzhyk commented on crates/dbsp/src/tutorial.rs:
```diff
@@ -1950,6 +1951,298 @@
+//!
+//! Finally, we point out that introducing a cycle to the graph prevents
+//! this fixed-point computation to stop terminating because then
+//! there is no fixed-point anymore. So pay attention to your data and
+//! your queries, if using more powerful (but somewhat more dangerous) iterative
+//! queries.
```
```suggestion
//! there is no fixed-point anymore.
```

I don't think that paying attention to data is the lesson here. The fix explained below is :)

ryzhyk reviewed (approved):
In my original review I missed the fact that this is using the single threaded API instead of `Runtime::init_circuit`. I think it's ok to merge as is and fix this later. But if you have the cycles to do it, please make that change too.

lstwn committed:
Use multi-threaded runtime for tutorial 10 and tutorial 11

The corresponding code blocks in the documentation still run single-threaded as the tutorial addresses runtime options later in the tutorial.

Signed-off-by: Leo Stewen <lstwn@mailbox.org>

lstwn committed:
Address minor feedback on the tutorial's text

Signed-off-by: Leo Stewen <lstwn@mailbox.org>

lstwn commented:
> In my original review I missed the fact that this is using the single threaded API instead of `Runtime::init_circuit`. I think it's ok to merge as is and fix this later. But if you have the cycles to do it, please make that change too.

I've switched to the multi-threaded runtime in [bd4ed5a](https://github.com/feldera/feldera/pull/3893/commits/bd4ed5a11a3ffcdec2de0ee00ee0a19f4e8f7760). But only for the code in the examples folder because the tutorial introduces DBSP's runtime at a later point. But note that all of your other example tutorial files still use the single-threaded `RootCircuit::build` API.
